eth-providers/src/base-provider.ts

```diff
@@ -1572,29 +1571,34 @@ export abstract class BaseProvider extends AbstractProvider {
     ]);
   };
 
-  _getTXReceipt = async (txHash: string): Promise<TransactionReceipt | TransactionReceiptGQL> => {
-    if (await this._isTXPending(txHash)) {
-      const txFromNextBlock = await this._getTXReceiptFromNextBlock(txHash);
-      if (txFromNextBlock) return txFromNextBlock;
-    }
+  _getTXReceipt = async (txHash: string): Promise<TransactionReceipt | TransactionReceiptGQL | null> => {
+    // TODO: potentially these 3 can go in parallel
+    let res = null;
+    try {
+      if (await this._isTXPending(txHash)) {
+        const txFromNextBlock = await this._getTXReceiptFromNextBlock(txHash);
+        if (txFromNextBlock) return txFromNextBlock;
+      }
 
-    const txFromCache = await this._getTxReceiptFromCache(txHash);
-    if (txFromCache) return txFromCache;
+      const txFromCache = await this._getTxReceiptFromCache(txHash);
+      if (txFromCache) return txFromCache;
 
-    try {
       const txFromSubql = await this.subql?.getTxReceiptByHash(txHash);
-      return txFromSubql || logger.throwError(`transaction hash not found`, Logger.errors.UNKNOWN_ERROR, { txHash });
+      res = txFromSubql || null;
     } catch {
-      return logger.throwError(`transaction hash not found`, Logger.errors.UNKNOWN_ERROR, { txHash });
+      res = null;
     }
+
+    return res;
   };
 
   // Queries
   getTransaction = (txHash: string): Promise<TransactionResponse> =>
     throwNotImplemented('getTransaction (deprecated: please use getTransactionByHash)');
 
-  getTransactionByHash = async (txHash: string): Promise<TX> => {
+  getTransactionByHash = async (txHash: string): Promise<TX | null> => {
     const tx = await this._getTXReceipt(txHash);
+    if (!tx) return null;
```

Comment:
The tx maybe is pending, need to return the tx(some fields null).

Reply:
https://eth.wiki/json-rpc/API

> Returns
> Object - A transaction receipt object, or null when no receipt was found

Reply:
if tx is pending, we will wait for it to be included in the block, then return the full receipt (with a timeout), this has been our behavior since the very beginning. 

I am fine if we want to change the behavior, but this will be a big breaking change and should be discussed more by the team

Reply:
Oh I found an issue from long ago, we can keep track of it https://github.com/AcalaNetwork/bodhi.js/issues/125

Reply:
as far as I remember, `eth_getTransactionReceipt` is very heavily used by all of the tools, such as metamask, hardaht, truffle, etc.. but `eth_getTransactionByHash` is rarely used, so I think #125 is not urgent

Reply:
`eth_getTransactionByHash` have no `pending` tag, only one param.  I think we can do it now.